test/testutils/validators.go

```diff
@@ -52,34 +52,65 @@ func ExpectLeaderSetExist(ctx context.Context, lws *leaderworkerset.LeaderWorker
 	}, Timeout, Interval).Should(gomega.Equal(true))
 }
 
-func ExpectValidServices(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet) {
+func ExpectValidServices(ctx context.Context, k8sClient client.Client, leaderWorkerSet *leaderworkerset.LeaderWorkerSet, numHeadlessServices int) {
 	gomega.Eventually(func() (bool, error) {
-		var headlessService corev1.Service
-		if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+
+		// Always got the latest lws.
+		var lws leaderworkerset.LeaderWorkerSet
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: leaderWorkerSet.Name, Namespace: leaderWorkerSet.Namespace}, &lws); err != nil {
 			return false, err
 		}
-		// we expect exactly one service and it must be of the correct type.
-		if headlessService.ObjectMeta.Name != lws.Name {
-			return false, errors.New("service name mismatch")
-		}
-		if headlessService.ObjectMeta.Namespace != lws.Namespace {
-			return false, errors.New("service namespace mismatch")
+
+		var headlessService corev1.Service
+		var headlessServiceList corev1.ServiceList
+		if err := k8sClient.List(ctx, &headlessServiceList, client.InNamespace(lws.Namespace)); err != nil {
+			return false, err
 		}
-		if headlessService.Spec.ClusterIP != "None" {
-			return false, errors.New("service type mismatch")
+
+		if len(headlessServiceList.Items) != (numHeadlessServices) {
```

Comment:
I don't think so, or at least not consistently. It can work after updates are done (check 1 if shared or check lws.spec.replicas if not) but during rollingUpdate with MaxSurge the number of services won't have either value and there's no way to identify that unless we do if rollingUpdateInProgress then lws.spec.replicas+maxSurge+1

Reply:
let's keep it as this then